File: concepts/Refactoring/PHP/PHP Refactoring Best Practices/ChangePreventers/0BaseProject/BaseProject/src/main/java/com/onlineshop/handlers/SimpleCurrencyConverter.java
package com.onlineshop.handlers;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

import static java.net.http.HttpClient.newBuilder;

public class SimpleCurrencyConverter {

    private String currencyTo;

    public SimpleCurrencyConverter(String currencyTo){
        this.currencyTo = currencyTo;
    }

    public double convert(double price){
        if(currencyTo.equalsIgnoreCase("EUR")){
            return price * 0.9;
        } else if (currencyTo.equalsIgnoreCase("CAD")){
            return price* 1.35;
        } else {
            throw new IllegalArgumentException("Unrecognized currency : " + currencyTo);
        }
    }

    public double convertWithWebService(double price){

        String rates = getCurrencyRates("USD");
        if(currencyTo.equalsIgnoreCase("EUR")){
            return price * getRate(rates, "EUR");

        } else if (currencyTo.equalsIgnoreCase("CAD")){
            return price * getRate(rates, "CAD");
        } else {
            throw new IllegalArgumentException("Unrecognized currency : " + currencyTo);
        }
    }


    private double getRate(String rates, String currencyTo) {
        return Double.valueOf(rates.substring(rates.lastIndexOf(currencyTo)).substring(5,9));
    }

    public String getCurrencyRates(String baseCurrency){
        HttpClient httpClient = newBuilder().build();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://api.exchangeratesapi.io/latest?base=" + baseCurrency))
                .build();

        HttpResponse<String> response = null;
        try {
            response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (IOException | InterruptedException e) {
            e.printStackTrace();
        }
        return response.body();
    }


}

File: concepts/Refactoring/PHP/PHP Refactoring Best Practices/3ChangePreventers/0BaseProject/src/handlers/SimpleCurrencyConverter.php
<?php

namespace app\handlers;

use Exception;
use GuzzleHttp\Client;

class SimpleCurrencyConverter
{
  private $currencyTo;

  public function __construct($currencyTo)
  {
    $this->currencyTo = $currencyTo;
  }

  public function convert($price)
  {
    if ($this->currencyTo === "EUR") {
      return $price * 0.9;
    } else if ($this->currencyTo === "CAD") {
      return $price * 1.35;
    } else
      throw new Exception("Unrecognized currency: $this->currencyTo");
  }


  public function convertWithWebService($price)
  {
    $rates = $this->getCurrencyRate("USD");
    if ($this->currencyTo === "EUR") {
      return $price * $this->getRate($rates, "EUR");
    } else if ($this->currencyTo === "CAD") {
      return $price * $this->getRate($rates, "CAD");;
    } else
      throw new Exception("Unrecognized currency: $this->currencyTo");
  }
  private function getRate(string $rates, string $currencyTo)
  {
    $data = json_decode($rates);
    return $data->rates->$currencyTo;
  }

  private function getCurrencyRate(string $baseCurrency)
  {
    $client = new Client(['base_uri' => 'https://api.exchangeratesapi.io/']);
    $response = $client->request('GET', "latest?base=$baseCurrency");
    return $response->getBody();
  }

  //new requirement: print rates : sol1 & 2
  public function printRates(string $currencyTo)
  {
    $rates = $this->getCurrencyRate("USD");
    if ($currencyTo === "EUR") {
      echo $this->getRate($rates, "EUR");
    } else if ($currencyTo === "CAD") {
      echo $this->getRate($rates, "CAD");;
    } else
      throw new Exception("Unrecognized currency: $this->currencyTo");
  }
  public function printRates1(string $currencyTo)
  {
    $rates = $this->getCurrencyRate("USD");
    echo $this->getRate($rates, $currencyTo);

  }
}
